File: src/rac/mcp/budget.py
"""Per-response character budget and whole-item truncation for Guide (ADR-033).

Guide tool responses are pasted straight into an agent's context window — the
scarcest resource in a session. One oversized response can flood that window
and drown out the very grounding Guide exists to provide, so every response
passes through a single per-response character budget before it reaches the
wire.

The budget counts *characters of the serialized JSON*, not tokens or bytes:
a character count is stable across models and tokenizer versions, which keeps
truncation deterministic and reproducible (ADR-032, ADR-033). The default cap
is 10,000 characters, fixed once at server startup — there is no per-call
override and no session state.

Reduction happens only at whole-item boundaries — whole search matches, whole
``incoming`` relationship entries, or a character prefix of a content tail —
never mid-element and never mid-JSON. A reduced response is stamped with three
marker fields, always appended *last* so the pinned field order of the tool
shapes survives:

- ``"truncated": true``
- ``"omitted": <count>`` — entries dropped, or characters dropped for a content
  tail
- ``"hint": "..."`` — how to narrow the request

A complete response carries none of these keys (``truncated`` is absent, not
``false``). The marker names and their trailing placement are part of the
pinned tool output contract.

Every tool serializes through :func:`serialize`, so truncation has exactly one
home and cannot drift between tools (ADR-033).
"""

from __future__ import annotations

import json

# Default per-response character budget (ADR-033). Set at startup via
# ``build_server(..., budget=...)``; no CLI flag, no per-call override.
DEFAULT_BUDGET = 10_000

# Pinned marker field names. Part of the output contract, so they are constants
# rather than string literals scattered through the truncation paths.
MARKER_TRUNCATED = "truncated"
MARKER_OMITTED = "omitted"
MARKER_HINT = "hint"

# Narrowing hints. Each is complete prose that stands alone in a transcript
# without the surrounding request (design: Accessibility).
HINT_SEARCH = "Narrow the query or request a specific artifact ID."
HINT_RELATED = "Request the artifact directly, or narrow what you are changing."
HINT_CONTENT = "Request a more specific artifact, or read the file directly for the full content."
HINT_SUMMARY = (
    "The repository summary exceeds the response budget; raise the server "
    "budget to see the full overview."
)


def serialize(payload: dict, budget: int = DEFAULT_BUDGET) -> str:
    """Serialize ``payload`` to JSON within ``budget``, truncating if needed.

    A payload that already fits is serialized unchanged, with no marker. One
    that does not is reduced at whole-item boundaries (see :func:`_truncate`)
    and stamped with the markers before serialization. The result is always
    valid JSON and, once a whole-item reduction fits, never exceeds ``budget``.
    When even the bare envelope is over budget, the marked-but-empty payload is
    returned — a structurally valid over-budget response beats unparseable
    noise (ADR-033).
    """
    if len(_encode(payload)) <= budget:
        return _encode(payload)
    return _encode(_truncate(payload, budget))


def _encode(payload: dict) -> str:
    """Serialize a tool payload deterministically — the unit the budget caps.

    ``ensure_ascii=False`` keeps a multibyte character as one character (the
    budget is a character count). ``sort_keys`` is deliberately omitted: each
    tool emits its keys in the pinned contract order, and the budget measures
    that exact serialization.
    """
    return json.dumps(payload, ensure_ascii=False)


def _truncate(payload: dict, budget: int) -> dict:
    """Reduce ``payload`` to fit ``budget``, dispatched on its truncatable field.

    Each tool shape carries exactly one unbounded field:

    - ``matches`` (search_artifacts) — drop whole match entries from the tail.
    - ``incoming`` (get_related) — drop whole incoming entries from the tail.
    - ``content`` (get_artifact) — drop characters from the content tail.

    get_summary has no unbounded field; an over-budget summary is marked without
    dropping data (``omitted == 0``), because there is no boundary to cut.
    """
    if "matches" in payload:
        return _truncate_list(payload, "matches", budget, HINT_SEARCH)
    if "incoming" in payload:
        return _truncate_list(payload, "incoming", budget, HINT_RELATED)
    if "content" in payload:
        return _truncate_content(payload, budget)
    return _marked(payload, HINT_SUMMARY, omitted=0)


def _truncate_list(payload: dict, key: str, budget: int, hint: str) -> dict:
    """Drop whole entries from the tail of ``payload[key]`` until it fits.

    Entries are dropped from the tail only, so the kept prefix is identical for
    identical input (determinism, ADR-032). ``omitted`` counts the dropped
    entries. If even the empty-list envelope is over budget, the fully-omitted
    marked payload is returned.
    """
    items = list(payload[key])
    total = len(items)
    kept = items
    while kept:
        candidate = _marked({**payload, key: kept}, hint, omitted=total - len(kept))
        if len(_encode(candidate)) <= budget:
            return candidate
        kept = kept[:-1]
    return _marked({**payload, key: []}, hint, omitted=total)


def _truncate_content(payload: dict, budget: int) -> dict:
    """Keep the largest head prefix of ``payload['content']`` that fits ``budget``.

    Characters are dropped from the tail so the kept head is identical for
    identical input. The largest fitting prefix is found by binary search over
    the prefix length; ``omitted`` is the number of characters removed. A tiny
    budget collapses the content to the empty string, still marked.
    """
    content = payload["content"]
    total = len(content)
    lo, hi, best = 0, total, 0
    while lo <= hi:
        mid = (lo + hi) // 2
        candidate = _marked(
            {**payload, "content": content[:mid]}, HINT_CONTENT, omitted=total - mid
        )
        if len(_encode(candidate)) <= budget:
            best = mid
            lo = mid + 1
        else:
            hi = mid - 1
    return _marked({**payload, "content": content[:best]}, HINT_CONTENT, omitted=total - best)


def _marked(base: dict, hint: str, *, omitted: int) -> dict:
    """Copy ``base`` and append the three truncation markers, in pinned order.

    ``base`` already carries any reduced field (a shorter list or a content
    prefix) in its original position; assigning the markers to the fresh copy
    lands them last, so each tool shape's pinned field order is preserved.
    """
    marked = dict(base)
    marked[MARKER_TRUNCATED] = True
    marked[MARKER_OMITTED] = omitted
    marked[MARKER_HINT] = hint
    return marked

File: rust/rac-engine/src/budget.rs
//! Per-response character budget (ADR-033) — a port of `src/rac/mcp/budget.py`
//! (ORACLE-NEXT revision, which adds the `items` rule). Shared by the CLI
//! retrieve surface (`commands::cmd_retrieve`) and the rac-mcp server.
//!
//! The budget unit is CHARACTERS (Python `len` of the serialized string —
//! Unicode code points, not bytes) of the payload serialized as
//! `json.dumps(payload, ensure_ascii=False)` with DEFAULT separators — i.e.
//! `", "` / `": "` WITH spaces (`pyjson::dumps_compact`). The docstring in the
//! oracle says "no spaces"; the code does not pass `separators`, so the wire
//! truth is *with* spaces (PORT-CONTRACT.d/10 §2) — port the code, not the
//! comment.
//!
//! Rule order (first matching key wins): `matches` → `incoming` → `items` →
//! `content` → mark-only (summary). Two overrun behaviors are ported
//! bug-for-bug (PORT-CONTRACT.d/10 §6):
//! - `get_summary` has no truncatable field: an over-budget summary is
//!   marked (`truncated:true, omitted:0, HINT_SUMMARY`) but nothing is
//!   dropped — the payload stays over budget.
//! - `get_related` with `depth>1`: only `incoming` shrinks; `neighborhood`
//!   is not truncatable, so the response can massively exceed the budget
//!   while carrying the marker.

use crate::pyjson::dumps_compact;
use serde_json::{json, Map, Value};

pub const DEFAULT_BUDGET: i64 = 10_000;

pub const MARKER_TRUNCATED: &str = "truncated";
pub const MARKER_OMITTED: &str = "omitted";
pub const MARKER_HINT: &str = "hint";

pub const HINT_SEARCH: &str = "Narrow the query or request a specific artifact ID.";
pub const HINT_RELATED: &str = "Request the artifact directly, or narrow what you are changing.";
pub const HINT_CONTENT: &str =
    "Request a more specific artifact, or read the file directly for the full content.";
pub const HINT_SUMMARY: &str = "The repository summary exceeds the response budget; raise the \
server budget to see the full overview.";
pub const HINT_RETRIEVE: &str = "Lower top_k, raise the budget, or narrow the task.";

/// `len(text)` in Python — code points, not bytes.
pub fn char_len(s: &str) -> i64 {
    s.chars().count() as i64
}

/// `text[:stop]` with Python slice semantics (negative stop trims the tail;
/// the truncators only pass non-negative stops, but the retrieve excerpt
/// share can go negative).
pub fn py_slice_to(s: &str, stop: i64) -> String {
    let n = char_len(s);
    let stop = if stop < 0 { (n + stop).max(0) } else { stop.min(n) };
    s.chars().take(stop as usize).collect()
}

fn length(payload: &Value) -> i64 {
    char_len(&dumps_compact(payload))
}

/// `budget.serialize(payload, budget)`.
pub fn serialize(payload: &Value, budget: i64) -> String {
    let text = dumps_compact(payload);
    if char_len(&text) <= budget {
        return text;
    }
    dumps_compact(&truncate(payload, budget))
}

fn truncate(payload: &Value, budget: i64) -> Value {
    let obj = payload.as_object().expect("payload is an object");
    if obj.contains_key("matches") {
        return truncate_list(payload, "matches", budget, HINT_SEARCH);
    }
    if obj.contains_key("incoming") {
        return truncate_list(payload, "incoming", budget, HINT_RELATED);
    }
    if obj.contains_key("items") {
        return truncate_items(payload, budget);
    }
    if obj.contains_key("content") {
        return truncate_content(payload, budget);
    }
    // No truncatable field (get_summary): mark, drop nothing (overrun #1).
    let mut marked = obj.clone();
    marked.insert(MARKER_TRUNCATED.to_string(), json!(true));
    marked.insert(MARKER_OMITTED.to_string(), json!(0));
    marked.insert(MARKER_HINT.to_string(), json!(HINT_SUMMARY));
    Value::Object(marked)
}

/// A copy of `payload` with `key` replaced by `kept` and the marker added.
/// `IndexMap::insert` keeps an existing key's position, matching Python
/// dict-update semantics (a `truncated` key already present — the
/// `get_related` edge-overflow marker — is overwritten in place).
fn with_marker(payload: &Value, key: &str, kept: Vec<Value>, omitted: i64, hint: &str) -> Value {
    let mut marked: Map<String, Value> = payload.as_object().expect("object").clone();
    marked.insert(key.to_string(), Value::Array(kept));
    marked.insert(MARKER_TRUNCATED.to_string(), json!(true));
    marked.insert(MARKER_OMITTED.to_string(), json!(omitted));
    marked.insert(MARKER_HINT.to_string(), json!(hint));
    Value::Object(marked)
}

fn truncate_list(payload: &Value, key: &str, budget: i64, hint: &str) -> Value {
    let items: Vec<Value> = payload[key].as_array().cloned().unwrap_or_default();
    let total = items.len() as i64;
    let mut kept = items;
    while !kept.is_empty() {
        let candidate = with_marker(payload, key, kept.clone(), total - kept.len() as i64, hint);
        if length(&candidate) <= budget {
            return candidate;
        }
        kept.pop();
    }
    with_marker(payload, key, Vec::new(), total, hint)
}

fn truncate_content(payload: &Value, budget: i64) -> Value {
    let content = payload["content"].as_str().unwrap_or("").to_string();
    let total = char_len(&content);
    let with_content = |kept: String, omitted: i64| -> Value {
        let mut marked: Map<String, Value> = payload.as_object().expect("object").clone();
        marked.insert("content".to_string(), json!(kept));
        marked.insert(MARKER_TRUNCATED.to_string(), json!(true));
        marked.insert(MARKER_OMITTED.to_string(), json!(omitted));
        marked.insert(MARKER_HINT.to_string(), json!(HINT_CONTENT));
        Value::Object(marked)
    };
    let (mut lo, mut hi) = (0i64, total);
    let mut best = 0i64;
    while lo <= hi {
        let mid = (lo + hi).div_euclid(2);
        let candidate = with_content(py_slice_to(&content, mid), total - mid);
        if length(&candidate) <= budget {
            best = mid;
            lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }
    with_content(py_slice_to(&content, best), total - best)
}

/// The retrieve `items` rule (ADR-113): excerpt-first, then whole-item.
fn truncate_items(payload: &Value, budget: i64) -> Value {
    let items: Vec<Value> = payload["items"].as_array().cloned().unwrap_or_default();
    let total = items.len() as i64;
    let mut kept = items;
    while !kept.is_empty() {
        let omitted = total - kept.len() as i64;
        let candidate = with_marker(payload, "items", kept.clone(), omitted, HINT_RETRIEVE);
        if length(&candidate) <= budget {
            return candidate;
        }
        // Trim the last kept item's excerpt before dropping it entirely.
        let mut last = kept
            .last()
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default();
        let excerpt: String = last
            .get("excerpt")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        let (mut lo, mut hi) = (0i64, char_len(&excerpt));
        let mut best: Option<i64> = None;
        while lo <= hi {
            let mid = (lo + hi).div_euclid(2);
            last.insert("excerpt".to_string(), json!(py_slice_to(&excerpt, mid)));
            let mut trial_items: Vec<Value> = kept[..kept.len() - 1].to_vec();
            trial_items.push(Value::Object(last.clone()));
            let trial = with_marker(payload, "items", trial_items, omitted, HINT_RETRIEVE);
            if length(&trial) <= budget {
                best = Some(mid);
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        if let Some(best) = best {
            last.insert("excerpt".to_string(), json!(py_slice_to(&excerpt, best)));
            let mut final_items: Vec<Value> = kept[..kept.len() - 1].to_vec();
            final_items.push(Value::Object(last));
            return with_marker(payload, "items", final_items, omitted, HINT_RETRIEVE);
        }
        kept.pop();
    }
    with_marker(payload, "items", Vec::new(), total, HINT_RETRIEVE)
}
